### ISSUE TITLE
Transpose with Decimal data type.

### ISSUE THREAD

**Issue by mintmi01** [modin-project/modin#4441]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04.4 LTS
- **Modin version** (`modin.__version__`): 0.13.3
- **Python version**: Python 3.9.5
- **Code we can use to reproduce**:
```python
import modin.pandas as mpd
import numpy as np
from modin.config import BenchmarkMode
import pandas as pd
import time
import decimal

BenchmarkMode.put(True)

print("Generating Data...")
modin_int_df = mpd.DataFrame(np.random.randint(0, 2, size=(751, 100_000)))
modin_decimal_df = modin_int_df.applymap(
                    lambda x: decimal.Decimal(x)
                    )

pandas_int_df = pd.DataFrame(np.random.randint(0, 2, size=(751, 100_000)))
pandas_decimal_df = pandas_int_df.applymap(
                    lambda x: decimal.Decimal(x)
                    )

print("Starting pandas int transpose now...")
start = time.time()
pandas_df_t = pandas_int_df.T
print(f"pandas int transpose took: {time.time() - start}")
print("Starting pandas decimal transpose now...")
start = time.time()
pandas_df_t = pandas_decimal_df.T
print(f"pandas decimal transpose took: {time.time() - start}")
print("Starting modin int transpose now...")
start = time.time()
modin_df_t = modin_int_df.T
print(f"modin int transpose took: {time.time() - start}")
print("Starting modin decimal transpose now...")
start = time.time()
modin_df_t = modin_decimal_df.T
print(f"modin decimal transpose took: {time.time() - start}")
```

### Describe the problem

When transposing a Modin DataFrame using the python [Decimal](https://docs.python.org/3/library/decimal.html) data type, I'm seeing a significant slowdown compared to transposing within modin using int as data type. Also, this behavior is not seen within a standard Pandas Dataframe. Please see the code snippet above.
### Source code / logs

*Code Output:*
```shell
Generating Data...
Starting pandas int transpose now...
pandas int transpose took: 0.012765884399414062
Starting pandas decimal transpose now...
pandas decimal transpose took: 0.011740446090698242
Starting modin int transpose now...
modin int transpose took: 5.335895776748657
Starting modin decimal transpose now...
modin decimal transpose took: 217.01361513137817
```

**Comment by mvashishtha** [comment]
@mintmi01 thank you for reporting this issue! I get nearly the same results results on an AWS x1e.16xlarge instance with 64 vCPU and 2.0 terabytes of RAM with Amazon Linux 2:

```
Starting pandas int transpose now...
pandas int transpose took: 0.008970499038696289
Starting pandas decimal transpose now...
pandas decimal transpose took: 0.012469768524169922
Starting modin int transpose now...
modin int transpose took: 4.7968363761901855
Starting modin decimal transpose now...
modin decimal transpose took: 219.47677755355835
```

@RehanSD is taking a look.

**Comment by mintmi01** [comment]
Thank you! Let me know if I can help add any additional detail.

**Comment by RehanSD** [comment]
Hi @mintmi01 ! Thank you so much for opening this issue! I've been doing some digging, and it seems like there may be two factors at play here - the first is that serializing Decimal objects is a lot slower than other types, and the second is that the transpose seems to be happening on only one worker, instead of being distributed. We're continuing to dig into this, but just wanted to give you an update on progress so far!

**Comment by mintmi01** [comment]
Hi @RehanSD, I see, sounds good. Thank you for the update!

**Comment by RehanSD** [comment]
Hi @mintmi01 ! Thank you so much for your patience! I have some updates on the first and second slowdowns! The second issue was easier to resolve - turns out our benchmarking mode was causing computation to occur on one node instead of utilizing all resources - we've opened an issue here (https://github.com/modin-project/modin/issues/4449) and will be fixing that soon!

As for the second issue, I was able to see performance gains across Mac and Linux by registering a custom serializer with Ray for the `Decimal` object type (code at the end). I've included the numbers below for a toy example we were working with (while we wait for my permissions from AWS for the larger instance). It would be helpful if you could try out the code on your end as well, to verify that this helps alleviate the issue!

M: `pd.DataFrame(np.random.randint(0, 2, size=(100_000, 751))).applymap(lambda x: decimal.Decimal(x))`
weights: `pd.Series(np.random.randint(0, 1000, size=100_000)).apply(lambda x: decimal.Decimal(x))`
Mac Times (16 vCPUs Intel Core i9 and 16 GB of memory):
* Without serialization fix: 1min 14s
* With serialization fix: 45.5 s

Linux Times (m4.4xlarge with 4 vCPUs and 64 GiB of memory):
* Without serialization fix: 1min 13s
* With serialization fix: 48.5s

To add the serialization fix, just add this code right before the `M.T.multiply` line!
```python
import pickle
from decimal import Decimal
def custom_serializer(decimal_obj):
    return pickle.dumps(decimal_obj)

def custom_deserializer(pickled_dec):
    return pickle.loads(pickled_dec)

ray.util.register_serializer(Decimal, serializer=custom_serializer, deserializer=custom_deserializer)
```

**Comment by RehanSD** [comment]
Just a quick update - even after the serialization fix above, serialization/storing outputs seems to be the bulk of the time - the `ray.timeline()` trace below shows that the `store_outputs` portion of the task (the green bars) are the most numerous + longest.
<img width="1536" alt="image" src="https://user-images.githubusercontent.com/17012605/168170847-9ebda511-50a5-4352-8b2a-45d482abc596.png">

**Comment by mvashishtha** [comment]
@mintmi01  @RehanSD there are two items that came out of this issue. I'll close this one in favor the more specific tasks. I have:

- #4449 : benchmark mode should kick off normally deferred operations, including transposes, in parallel.
- #4506: We need better performance on Decimal data in general

@mintmi01 thank you again for reporting this!
